File: Python3/0589-N-ary-Tree-Preorder-Traversal/soln-1.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
class Solution(object):
    def preorder(self, root):
        """
        :type root: Node
        :rtype: List[int]
        """
        if not root:
            return []
        stack, vals = [root], []
        while stack:
            node = stack.pop()
            vals.append(node.val)
            for child in reversed(node.children):
                stack.append(child)
        return vals

File: C++/0783-Minimum-Distance-Between-BST-Nodes/soln-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDiffInBST(TreeNode* root) {
        vector<int> vals;
        Preorder(root, & vals);
        int ans = vals[1] - vals[0];
        for(int i = 1; i < (int) vals.size(); ++i) {
            ans = min(ans, vals[i] - vals[i - 1]);
        }
        return ans;
    }
private:
    void Preorder(TreeNode * node, vector<int> * vals) {
        if(node != nullptr) {
            Preorder(node->left, vals);
            vals->push_back(node->val);
            Preorder(node->right, vals);
        }
    }
};
